src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2488,41 +2579,71 @@ var projectUniqueNamesForBuildIntToUpdate
+                {
+                    { buildIntegratedProject.MSBuildProjectPath, nuGetProjectActions}
+                },
+                packageIdentity: null, // since we have nuGetProjectActions no need packageIdentity
+                primarySources: null, // since we have nuGetProjectActions no need primarySources
+                nuGetProjectContext,
+                token
+                );
 
-            // Add all enabled sources for the existing packages
-            var enabledSources = SourceRepositoryProvider.GetRepositories();
+            return resolvedAction.FirstOrDefault(r => r.Project == buildIntegratedProject)?.Action as BuildIntegratedProjectAction;
+        }
 
-            sources.UnionWith(enabledSources);
 
-            // Read the current lock file if it exists
-            LockFile originalLockFile = null;
-            var lockFileFormat = new LockFileFormat();
+        /// <summary>
+        /// Run project actions for build integrated many projects.
+        /// </summary>
+        internal async Task<IEnumerable<ResolvedAction>> PreviewBuildIntegratedProjectsActionsAsync(
+            IEnumerable<BuildIntegratedNuGetProject> buildIntegratedProjects,
+            Dictionary<string, IEnumerable<NuGetProjectAction>> nugetProjectActionsLookup,
+            PackageIdentity packageIdentity,
+            IEnumerable<SourceRepository> primarySources,
+            INuGetProjectContext nuGetProjectContext,
+            CancellationToken token)
+        {
+            if (nugetProjectActionsLookup == null)
+            {
+                nugetProjectActionsLookup = new Dictionary<string, IEnumerable<NuGetProjectAction>>(PathUtility.GetStringComparerBasedOnOS());
+            }
 
-            var lockFilePath = await buildIntegratedProject.GetAssetsFilePathAsync();
+            if (buildIntegratedProjects == null)
+            {
+                throw new ArgumentNullException(nameof(buildIntegratedProjects));
+            }
 
-            if (File.Exists(lockFilePath))
+            if (!buildIntegratedProjects.Any())
             {
-                originalLockFile = lockFileFormat.Read(lockFilePath);
+                // Return empty if there are no buildIntegratedProjects.
+                return Enumerable.Empty<ResolvedAction>();
             }
 
-            var logger = new ProjectContextLogger(nuGetProjectContext);
-            var dependencyGraphContext = new DependencyGraphCacheContext(logger, Settings);
+            if (nuGetProjectContext == null)
+            {
+                throw new ArgumentNullException(nameof(nuGetProjectContext));
+            }
 
-            // Get Package Spec as json object
-            var originalPackageSpec = await DependencyGraphRestoreUtility.GetProjectSpec(buildIntegratedProject, dependencyGraphContext);
+            if (!nugetProjectActionsLookup.Any() && packageIdentity == null)
```

Comment:
Is packageIdentity != null && nugetProjectActionsLookup.Any()  a valid state?

Reply:
Currently this scenario not expected input.
We need either valid  `NugetProjectAction` vs  `packageIdentity & activeSource`, but not both.
For example here we pass both `packageIdentity & activeSource`, but `NugetProjectAction` = null:
https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1598-L1600

Here we pass only `NugetProjectAction`, but not  `packageIdentity & activeSource`:
https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2584-L2589

Just in case I do more validation on [Line#2668] to find out any bad input parameter: (https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2668) and [Line#2679](https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2679).